src/dlr_tensorflow/dlr_tensorflow.cc

```diff
@@ -70,73 +72,90 @@ void TensorflowModel::LoadFrozenModel(const char* pb_file) {
+    std::vector<int64_t> shape;
+    for (auto& v : k.second.array_items()) {
+      shape.push_back(v.int_value());
     }
+    input_shapes_.push_back(shape);
+  }
+  num_inputs_ = input_names_.size();
+}
 
-    TF_Operation* op = TF_GraphOperationByName(graph_, op_name.c_str());
-    if (op == NULL) {
-      LOG(FATAL)
-          << "ERROR: inputOp TF_GraphOperationByName failed for operation "
-          << op_name;
-      return;  // unreachable
-    }
-    TF_Output oper_out = {op, op_out_id};
-    const int n_dim = TF_GraphGetTensorNumDims(graph_, oper_out, status_);
-    if (TF_GetCode(status_) != TF_OK) {
-      LOG(FATAL) << "ERROR: TF_GraphGetTensorNumDims failed "
-                 << TF_Message(status_);
-      return;  // unreachable
-    }
-    int64_t dims[n_dim];
-    TF_GraphGetTensorShape(graph_, oper_out, dims, n_dim, status_);
+void TensorflowModel::ParseOutputSpec(const std::string& output_arr) {
+  std::string err;
+  const auto json = json11::Json::parse(output_arr, err);
+  if (!err.empty()) {
+    LOG(FATAL) << "ERROR: Failed to parse output_arr: " << err;
+    return;  // unreachable
+  }
+  for (auto& k : json.array_items()) {
+    output_names_.push_back(k.string_value());
+  }
+  num_outputs_ = output_names_.size();
+}
+
+TF_Output TensorflowModel::ParseTensorName(const std::string& t_name) {
+  std::regex r("^(.+):(\\d+)$");
+  std::smatch match;
+  if (!std::regex_search(t_name, match, r)) {
+    LOG(FATAL) << "ERROR: failed to parse tensor name " << t_name;
+  }
+  std::string op_name = match.str(1);
+  int op_out_id = std::stoi(match.str(2));
+  TF_Operation* op = TF_GraphOperationByName(graph_, op_name.c_str());
+  if (op == NULL) {
+    LOG(FATAL) << "ERROR: TF_GraphOperationByName failed for operation "
+               << op_name;
+  }
+  TF_Output oper_out = {op, op_out_id};
+  return oper_out;
+}
+
+void TensorflowModel::PrepInputs() {
+  for (int i = 0; i < num_inputs_; i++) {
+    std::string t_name = input_names_[i];
+    TF_Output oper_out = ParseTensorName(t_name);
+    std::vector<int64_t> shape = input_shapes_[i];
+    int64_t* dims = shape.data();
+    size_t n_dim = shape.size();
+    TF_GraphSetTensorShape(graph_, oper_out, dims, n_dim, status_);
     if (TF_GetCode(status_) != TF_OK) {
-      LOG(FATAL) << "ERROR: TF_GraphGetTensorShape failed "
+      LOG(FATAL) << "ERROR: TF_GraphSetTensorShape failed "
                  << TF_Message(status_);
       return;  // unreachable
     }
-    // Set fixed batch size if batch size is dynamic
-    if (dims[0] == -1) {
-      dims[0] = batch_size > 0 ? batch_size : 1;
-      TF_GraphSetTensorShape(graph_, oper_out, dims, n_dim, status_);
-      if (TF_GetCode(status_) != TF_OK) {
-        LOG(FATAL) << "ERROR: TF_GraphSetTensorShape failed "
-                   << TF_Message(status_);
-        return;  // unreachable
-      }
-    }
+
     size_t num_elements = 1;
     for (int z = 0; z < n_dim; z++) {
-      if (dims[z] == -1) {
```

Comment:
Should we keep this check in case user gives a bad input shape containing -1?

Reply:
good catch!